File: Лабороторная работа 1/tennis.cpp
#include "tennis.h"
#include <iostream>
using namespace std;

// Конструктор 1: полный
Tennis::Tennis(string name, int players, string surface, int sets, int score1, int score2, int game)
    : SportEvent(name, players), courtSurface(surface), setsCount(sets), 
      player1Score(score1), player2Score(score2), currentGame(game) 
{}

// Конструктор 2: без счета и гейма
Tennis::Tennis(string name, int players, string surface, int sets)
    : SportEvent(name, players), courtSurface(surface), setsCount(sets), 
      player1Score(0), player2Score(0), currentGame(1) 
{}

// Конструктор 3: минимальный
Tennis::Tennis(string name, int players)
    : SportEvent(name, players), courtSurface("Грунт"), setsCount(3), 
      player1Score(0), player2Score(0), currentGame(1) 
{}

void Tennis::startGame() {
    isActive = true;
    cout << eventName << " (теннис) начался!" << endl;
    cout << "  Покрытие корта: " << courtSurface << endl;
    cout << "  Матч до " << setsCount << " сетов" << endl;
    cout << "  Текущий счет: " << player1Score << " - " << player2Score 
         << " (гейм " << currentGame << ")" << endl;
}

void Tennis::stopGame() {
    isActive = false;
    cout << eventName << " (теннис) завершен. Финальный счет: " 
         << player1Score << " - " << player2Score << endl;
}

void Tennis::showInfo() {
    cout << "Событие: " << eventName 
         << " | Вид спорта: Теннис"
         << " | Статус: " << (isActive ? "ИДЕТ МАТЧ" : "НЕ АКТИВЕН")
         << " | Покрытие: " << courtSurface
         << " | Сетов: " << setsCount
         << " | Счет: " << player1Score << " - " << player2Score
         << " (гейм " << currentGame << ")" << endl;
}

void Tennis::serve(string player) {
    if (isActive) {
        cout << eventName << ": Матч приостановлен и идёт выполнение подачи" << endl;
         cout << eventName << ": " << player << " выполняет подачу. ";
        return;
    }
    
    
    // Симуляция подачи (70% успешных)
    bool success = (rand() % 100) < 70;
    if (success) {
        cout << "Подача успешна!" << endl;
        
        // Начисление очков (40% вероятность выигрыша очка)
        bool winPoint = (rand() % 100) < 40;
        if (winPoint) {
            if (player == "Игрок 1") {
                player1Score++;
                cout << "  " << player << " выигрывает очко! Счет: " 
                     << player1Score << " - " << player2Score << endl;
            } else if (player == "Игрок 2") {
                player2Score++;
                cout << "  " << player << " выигрывает очко! Счет: " 
                     << player1Score << " - " << player2Score << endl;
            }
        }
    } else {
        cout << "Ошибка подачи!" << endl;
    }
}

void Tennis::changeSides() {
    if (isActive) {
        cout << eventName << ": Матч приостановлен и идёт смена сторон" << endl;
        cout << eventName << ": Смена сторон. Игроки меняются сторонами корта." << endl;
        currentGame++;
        return;
    }
    
    cout << eventName << ": Смена сторон. Игроки меняются сторонами корта." << endl;
    currentGame++;
}

File: Лабороторная работа 2/Models/Tennis.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace SportEventApp.Models
{
    [Table("TennisEvents")]
    public class Tennis : SportEvent
    {
        [MaxLength(50)]
        public string CourtSurface { get; set; } = string.Empty;
        
        public int SetsCount { get; set; }
        public int Player1Score { get; set; }
        public int Player2Score { get; set; }
        public int CurrentGame { get; set; }
        
        private static Random _random = new Random();
        
        public Tennis() : base() { }
        
        public Tennis(string name, int players, string surface, int sets, int score1, int score2, int game)
            : base(name, players)
        {
            CourtSurface = surface;
            SetsCount = sets;
            Player1Score = score1;
            Player2Score = score2;
            CurrentGame = game;
            EventType = "Tennis";
        }
        
        public Tennis(string name, int players, string surface, int sets)
            : this(name, players, surface, sets, 0, 0, 1) { }
        
        public Tennis(string name, int players)
            : this(name, players, "Грунт", 3, 0, 0, 1) { }
        
        public override void StartGame()
        {
            IsActive = true;
        }
        
        public override void StopGame()
        {
            IsActive = false;
        }
        
        public override string ShowInfo()
        {
            return $"Событие: {EventName} | Вид: Теннис | Статус: {(IsActive ? "ИДЕТ МАТЧ" : "НЕ АКТИВЕН")} | Покрытие: {CourtSurface} | Сетов: {SetsCount} | Счет: {Player1Score}-{Player2Score} (гейм {CurrentGame})";
        }
        
        public string Serve(string player)
        {
            if (!IsActive)
                return $"{EventName}: Сначала начните матч!";
            
            bool success = _random.Next(100) < 70;
            
            if (success)
            {
                bool winPoint = _random.Next(100) < 40;
                if (winPoint)
                {
                    if (player == "Игрок 1")
                    {
                        Player1Score++;
                        return $"{EventName}: {player} выполняет подачу. Подача успешна! {player} выигрывает очко! Счет: {Player1Score}-{Player2Score}";
                    }
                    else if (player == "Игрок 2")
                    {
                        Player2Score++;
                        return $"{EventName}: {player} выполняет подачу. Подача успешна! {player} выигрывает очко! Счет: {Player1Score}-{Player2Score}";
                    }
                }
                return $"{EventName}: {player} выполняет подачу. Подача успешна!";
            }
            return $"{EventName}: {player} выполняет подачу. Ошибка подачи!";
        }
        
        public string ChangeSides()
        {
            CurrentGame++;
            return $"{EventName}: Смена сторон. Игроки меняются сторонами корта. Текущий гейм: {CurrentGame}";
        }
        
        public string GetCourtSurface() => CourtSurface;
        public void SetCourtSurface(string surface) => CourtSurface = surface;
        
        public int GetSetsCount() => SetsCount;
        public void SetSetsCount(int sets) => SetsCount = sets;
        
        public int GetPlayer1Score() => Player1Score;
        public void SetPlayer1Score(int score) => Player1Score = score;
        
        public int GetPlayer2Score() => Player2Score;
        public void SetPlayer2Score(int score) => Player2Score = score;
        
        public int GetCurrentGame() => CurrentGame;
        public void SetCurrentGame(int game) => CurrentGame = game;
    }
}